### PR TITLE
[SPARK-14346] SHOW CREATE TABLE for data source tables

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

This PR adds native `SHOW CREATE TABLE` DDL command for data source tables. Support for Hive tables will be added in follow-up PR(s).

To show table creation DDL for data source tables created by CTAS statements, this PR also added partitioning and bucketing support for normal `CREATE TABLE ... USING ...` syntax.
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

A new test suite `ShowCreateTableSuite` is added in sql/hive package to test the new feature.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 1]
```diff
@@ -117,101 +112,3 @@ case class ExplainCommand(
...
-            s"Non-partitioning column(s) [${badColumns.mkString(", ")}] are " +
-              s"specified for SHOW PARTITIONS")
-        }
-      }
-      val partNames =
-        catalog.listPartitions(table, spec).map(p => getPartName(p.spec, tab.partitionColumnNames))
-      partNames.map { p => Row(p) }
-    }
-  }
-}
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Moved to `tables.scala`, not related to this PR.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -45,7 +45,9 @@ statement
     | ALTER DATABASE identifier SET DBPROPERTIES tablePropertyList     #setDatabaseProperties
     | DROP DATABASE (IF EXISTS)? identifier (RESTRICT | CASCADE)?      #dropDatabase
     | createTableHeader ('(' colTypeList ')')? tableProvider
-        (OPTIONS tablePropertyList)?                                   #createTableUsing
+        (OPTIONS tablePropertyList)?
+        (PARTITIONED BY partitionColumnNames=identifierList)?
+        bucketSpec?                                                    #createTableUsing
```
Comment (yhuai) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @cloud-fan Do we allow users to specify bucketing without providing partitioning columns? Seems only `DynamicPartitionWriterContainer` support bucketSpec?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 3]
```diff
@@ -389,3 +392,238 @@ case class ShowTablePropertiesCommand(table: TableIdentifier, propertyKey: Optio
...
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+
+    if (catalog.isTemporaryTable(table)) {
+      throw new AnalysisException(
+        s"SHOW CREATE TABLE cannot be applied to temporary table")
+    }
+
+    require(catalog.tableExists(table), s"Table $table doesn't exist")
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Can we use AnalysisException?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 4]
```diff
@@ -389,3 +392,238 @@ case class ShowTablePropertiesCommand(table: TableIdentifier, propertyKey: Optio
...
+    val tableMetadata = catalog.getTableMetadata(table)
+
+    val stmt = if (tableMetadata.properties.contains("spark.sql.sources.provider")) {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Use DDLUtils.isDatasourceTable?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Thanks, this is neat.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 3 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Do we tend to use `AnalysisException` in all use cases where catalog metadata is involved? I ask this because, strictly speaking, the method above isn't invoked during analysis phase.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Yes, we do. If partition columns are empty but bucket columns are not, we will also use `DynamicPartitionWriterContainer`, see: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelation.scala#L124-L136

Comment (yhuai) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 5]
```diff
@@ -452,3 +455,241 @@ case class ShowTablePropertiesCommand(table: TableIdentifier, propertyKey: Optio
...
+      )
+    )
+
+    if (schemaParts.nonEmpty) {
+      val fields = DataType.fromJson(schemaParts.mkString).asInstanceOf[StructType].fields
+      val colTypeList = fields.map(f => s"${quoteIdentifier(f.name)} ${f.dataType.sql}")
+      builder ++= colTypeList.mkString("(", ", ", ")")
+    }
+
+    builder ++= "\n"
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: I think having a utility function in DDLUtils to construct the DataType from those strings stored in table properties will be very useful. This function can also help us to provide a nice output of describe table (when the table is a data source table). We can do it in a separate PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 6]
```diff
@@ -452,3 +455,241 @@ case class ShowTablePropertiesCommand(table: TableIdentifier, propertyKey: Optio
...
+        throw new AnalysisException(
+          s"Corrupted $typeName in catalog: $numCols parts expected, but part $index is missing."
+        )
+      )).map(quoteIdentifier)
+    }
+
+    val partCols = getColumnNamesByType("part", "partitioning columns")
+    if (partCols.nonEmpty) {
+      builder ++= s"PARTITIONED BY ${partCols.mkString("(", ", ", ")")}\n"
+    }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: We should also have utility functions for reconstructing partition columns, bucketing columns, and sorting columns from strings.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    val expected = sqlContext.externalCatalog.getTable(db, table.table)
+    val shownDDL = sql(s"SHOW CREATE TABLE ${table.quotedString}").head().getString(0)
+    val newTableName = s"${table.table}_new"
+
+    withTable(newTableName) {
+      assertResult(expected) {
+        val newDDL = shownDDL.replaceFirst(table.table, newTableName)
+        sql(newDDL)
+        val newTable = sqlContext.externalCatalog.getTable(db, table.table)
+        newTable.copy(identifier = table)
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: `createTime` in `CatalogTable` did not introduce any issue?

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: Good question... I start to wonder why the test ever passed... Looking into it.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 8]
```diff
@@ -45,7 +45,9 @@ statement
...
     | createTableHeader tableProvider
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: not related to this PR, but should we combine these 2 rules? They are all called `createTableUsing` but the second one is actually `createTableUsingAsSelect`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: I think we should also add a TODO here. Putting these informations in table property is a workaround for hive. After we have a better isolation of hive, we should be able to hide this logic in hive external catalog and don't need to do this trick here.

**Comment by cloud-fan** [comment]
LGTM except the weird test-should-fail issue.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala (hunk 7 above)
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: Err, turned out to be a stupid error: I put `table.table` instead of `newTableName` when retrieving the newly created table...

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: ah i see. So, we need to update this test?

**Comment by liancheng** [comment]
Updated test suite by normalizing `CatalogTable`s before comparing them.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 8 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: The reason for not combining them is that the first is a `CREATE TABLE` statement (and has column and no query) and that the second one is `CREATE TABLE ... AS` statement (which has no columns and a query). This way the parser will catch input that both defines columns and a query.

We could also check this in the `SparkSqlAstBuilder`...

**Comment by liancheng** [comment]
Part of the changes can be simplified using DDL utility methods introduced in #13025.

**Comment by cloud-fan** [comment]
#13025 is merged, are you going to update this PR?

**Comment by liancheng** [comment]
I think we can merge this one first. There are some other places where the new DDL utility methods can be used to simplify code. I can fix them altogether in follow-up PRs.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  private def checkCreateTable(table: TableIdentifier): Unit = {
+    val db = table.database.getOrElse("default")
+    val expected = spark.externalCatalog.getTable(db, table.table)
+    val shownDDL = sql(s"SHOW CREATE TABLE ${table.quotedString}").head().getString(0)
+    sql(s"DROP TABLE ${table.quotedString}")
+
+    withTable(table.table) {
+      val newDDL = shownDDL.replaceFirst(table.table, table.table)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: not needed?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        "numRows",
+        "rawDataSize",
+        "totalSize",
+        "totalNumberFiles",
+        "maxFileSize",
+        "minFileSize"
+      )
+
+      table.copy(
+        identifier = expected.identifier,
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: looks like `expected` and `actual` will always have same identifier? We remove the table and create a new one with same name in `checkCreateTable`.

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: Yea, thanks, this is another place that I forgot to remove after simplifying the test infra.

**Comment by cloud-fan** [comment]
LGTM

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
The last build failure was a compilation failure in a file not even touched by this PR. Trying again.

**Comment by yhuai** [comment]
yea, build was broken

**Comment by yhuai** [comment]
test this please

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (yhuai) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I just realized that `OPTIONS` goes first. I am wondering if it makes sense putting `OPTIONS` after `PARTITIONED BY` and bucketSpec.

Comment (yhuai) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Let's add tests for this syntax (right now we only have tests for show create table).

**Comment by yhuai** [comment]
Thanks. I am going to merge this to master and branch 2.0.

**Comment by yhuai** [comment]
Let's have a follow-up prs to make the following changes:
1. Make the command use those new helper functions added in DDLUtils.
2. Add tests for new keywords added in CREATE TABLE USING syntax
3. Support tables stored with Hive formats.
